### PR TITLE
feat: Add numba harmonic mean implementation

### PR DESCRIPTION
**Pull request by lsetiawan**
Added '_sv_harmon_mean' numba implementation for
calculating harmonic mean that uses calculations
used within the original fortran code to get same
harmonic values. Now there's a 'method' value for the main 'sv_harmonic_mean' function that allows user
to switch back and forth b/w the two methods.

---

Ref #193

### CODE REVIEW

File: src/gnatss/harmonic_mean.py [hunk 1]
```diff
@@ -44,7 +45,77 @@ def _compute_hm(svdf: pd.DataFrame, start_depth: float, end_depth: float) -> flo
...
+
+        zf = z2
+        if z2 < ze:
+            zf = ze
+
+        # Compute slope
+        b = (c_z2 - c_z1) / (z2 - z1)
+
+        # Ensure that slope is not 0
+        assert b != 0.0
```
Comment (carlosgjs) [inline src/gnatss/harmonic_mean.py:101]:
```suggestion
        assert b != 0.0, "Slope is zero"
```

File: src/gnatss/harmonic_mean.py [hunk 2]
```diff
@@ -44,7 +45,77 @@ def _compute_hm(svdf: pd.DataFrame, start_depth: float, end_depth: float) -> flo
...
+        The end depth for harmonic mean to be computed
+
+    Returns
+    -------
+    float
+        The sound speed harmonic mean result
+    """
+    # Ensure that depth and sound speed arrays
+    # are the same shape
+    assert dd.shape == sv.shape
```
Comment (carlosgjs) [inline src/gnatss/harmonic_mean.py:79]:
```suggestion
    assert dd.shape == sv.shape, f"dd and sv should have the same shape. dd:{dd.shape} != sv:{sv.shape}"
```

File: src/gnatss/harmonic_mean.py [hunk 3]
```diff
@@ -67,9 +145,35 @@ def sv_harmonic_mean(svdf: pd.DataFrame, start_depth: float, end_depth: float) -
...
+        abs_sv = abs(svdf)
+
+        svhm = _compute_hm(abs_sv, abs_start, abs_end)
+    elif method == "numba":
+        # Ensure both start and end depths are negative
+        start_depth = -np.abs(start_depth)
+        end_depth = -np.abs(end_depth)
+
+        if start_depth < end_depth:
+            raise ValueError("Start depth must be greater than end depth!")
```
Comment (carlosgjs) [inline src/gnatss/harmonic_mean.py:166]: nit: Include the actual values in the error for easier debugging

File: src/gnatss/harmonic_mean.py [hunk 4]
```diff
@@ -67,9 +145,35 @@ def sv_harmonic_mean(svdf: pd.DataFrame, start_depth: float, end_depth: float) -
...
+
+        svdf = svdf[(svdf[SP_DEPTH].round() <= start_depth)]
+        # Extract the numpy arrays for depth and sound speed
+        # Make sure that the depth array has negative values
+        # and the sound speed array has positive values
+        dd = -np.abs(svdf[SP_DEPTH].values)
+        sv = np.abs(svdf[SP_SOUND_SPEED].values)
+
+        svhm = _sv_harmon_mean(dd, sv, start_depth, end_depth)
+    else:
```
Comment (carlosgjs) [inline src/gnatss/harmonic_mean.py:176]: Question for my own understanding.
Who enforces that only the valid literal values are used? Is that mypy or a runtime check. Ie. at runtime could we actually get here?

**Review by carlosgjs — APPROVED** [review]
LGTM. Some inline questions, sugggestions.

File: src/gnatss/harmonic_mean.py (hunk 4 above)
Comment (lsetiawan) [reply src/gnatss/harmonic_mean.py:176]: If the `method` value is neither a "scipy" or "numba", some other string would go to the else statement. Currently a runtime check happens but there's typehints and documentation of the valid values.

File: src/gnatss/harmonic_mean.py [hunk 5]
```diff
@@ -67,9 +145,35 @@ def sv_harmonic_mean(svdf: pd.DataFrame, start_depth: float, end_depth: float) -
...
-    # Make all of the values absolute values, so we're only dealing with positives
-    abs_start = abs(start_depth)
-    abs_end = abs(end_depth)
-    abs_sv = abs(svdf)
 
-    return _compute_hm(abs_sv, abs_start, abs_end)
+    for col in [SP_DEPTH, SP_SOUND_SPEED]:
+        if col not in svdf.columns:
+            raise ValueError(f"{col} column must exist in the input dataframe!")
+
```
Comment (lsetiawan) [inline src/gnatss/harmonic_mean.py:152]:
```suggestion

    method = method.lower()
```

File: src/gnatss/harmonic_mean.py (hunk 3 above)
Comment (lsetiawan) [inline src/gnatss/harmonic_mean.py:166]:
```suggestion
            raise ValueError(f"Start depth {start_depth} must be greater than end depth {end_depth}!")
```

**Pull request merged**